src/redprl/refiner_types.fun

```diff
@@ -53,169 +53,170 @@ struct
+          let
+            val (psi, pathty) = synthTerm' sign tr H (m, m)
+            val Syn.PATH ((x, a), left, right) = Syn.out pathty
+            val n =
+              case Syn.out r of
+                 Syn.DIM0 => left
+               | Syn.DIM1 => right
+            val ty = substVar (r, x) a
+            val goal = makeEq tr H ((tm1, n), ty)
+          in
+            (goal :: psi, ty)
+          end
+
+        | (O.PROJ lbl $ [_ \ m], O.PROJ lbl' $ [_ \ n]) =>
+          let
+            val true = lbl = lbl'
+            val (psi, rcdty) = synthTerm' sign tr H (m, n)
+            val Abt.$ (O.RECORD lbls, args) = out rcdty
+
+            val i = #1 (Option.valOf (ListUtil.findEqIndex lbl lbls))
+            val (us \ ty) = List.nth (args, i)
+
+            (* supply the dependencies *)
+            val lblPrefix = List.take (lbls, i)
+            val rho = ListPair.mapEq (fn (lbl, u) => (Syn.into @@ Syn.PROJ (lbl, m), u)) (lblPrefix, us)
+            val ty = VarKit.substMany rho ty
+          in
+            (psi, ty)
+          end
+
+        | (O.S1_REC $ [[x] \ cx, _ \ m, _, _], O.S1_REC $ _) =>
+          let
+            val ty = substVar (m, x) cx
+            val goal = makeEq tr H ((tm1, tm2), ty)
+          in
+            ([goal], ty)
+          end
+
+        | (O.IF $ [[x] \ cx, _ \ m, _, _], O.IF $ _) =>
+          let
+            val ty = substVar (m, x) cx
+            val goal = makeEq tr H ((tm1, tm2), ty)
+          in
+            ([goal], ty)
+          end
+
+        | (O.PUSHOUT_REC $ [[x] \ cx, _ \ m, _, _, _], O.PUSHOUT_REC $ _) =>
+          let
+            val ty = substVar (m, x) cx
+            val goal = makeEq tr H ((tm1, tm2), ty)
+          in
+            ([goal], ty)
+          end
+
+        | (O.COEQUALIZER_REC $ [[x] \ cx, _ \ m, _, _], O.COEQUALIZER_REC $ _) =>
+          let
+            val ty = substVar (m, x) cx
+            val goal = makeEq tr H ((tm1, tm2), ty)
+          in
+            ([goal], ty)
+          end
+
+        | (O.NAT_REC $ [[x] \ cx, _ \ m, _, _], O.NAT_REC $ _) =>
+          let
+            val ty = substVar (m, x) cx
+            val goal = makeEq tr H ((tm1, tm2), ty)
+          in
+            ([goal], ty)
+          end
+        | (O.INT_REC $ [[x] \ cx, _ \ m, _, _, _, _], O.INT_REC $ _) =>
+          let
+            val ty = substVar (m, x) cx
+            val goal = makeEq tr H ((tm1, tm2), ty)
+          in
+            ([goal], ty)
+          end
+
+    and synthTerm' sign tr H (tm1, tm2) =
+      let
+        val (psi, ty) = synthNeutral' sign tr H (Machine.eval sign Machine.STABLE Machine.Unfolding.never tm1, Machine.eval sign Machine.STABLE Machine.Unfolding.never tm2)
+      in
+        (psi, Machine.eval sign Machine.STABLE Machine.Unfolding.always ty)
+      end
+
+    fun synthTerm sign tr H (tm1, tm2) =
+      let
+        val (psi, ty) = synthTerm' sign tr H (tm1, tm2)
+      in
+        (List.rev psi, ty)
+      end
   end
 
+
+
+
+
+
```

Comment:
Just curious... are there secret messages hidden in these lines?

Reply:
lol, I can delete them 😉